### ISSUE TITLE
Shoko strange behavior for corrupted files replaced with good ones

### ISSUE THREAD

**Issue by bigretromike** [japanesemediamanager/ShokoServer#724]
While adding collection,
few files got corrupted
shoko hash them 
wrong hashes go inside db
adding new file (proper, not corrupted) into new import folder result in being added with correct CRC hash but bad HASH resulting with being unable to recognize file by anidb
The only solution to make files being recognize correctly is via ShokoDesktop > Utilites > Unrecognized > Rehash All, because using Server UI don't fix it
(tried, removing and moving with "remove missing files" from ui and importing new folder)

**Comment by Cazzar** [comment]
It's a matter of when the filesystem detects the copy and when we start
hashing,
If you have copied the files in, and we hash half way through the copy, it
will get a collision.
Though tbh with CRC32 it's easy to collide

**Comment by bigretromike** [comment]
its not about copy;
file was corrupted - same size, wrong hash.
shoko scanned it and added to db as unrecognized (correct) with crc that was not official released (corrupted)
then while trying to replace it shoko hashed it again, but only crc was rehashed (correct crc hash) but the "hash" that is send to anidb wasn't rehashed again and it was same as the one from corrupted file;

**Comment by Cazzar** [comment]
Maybe I could also be saying it can occur then too?

**Comment by bigretromike** [comment]
If I recall there was talk years ago that it wont hash file that are in middle on transfering because some system lock etc... but I don't know how accurate is that and if it's Other OS wide. So maybe sure, I just had this with corrupted files.

**Comment by maxpiva** [comment]
Please, check your memory with memtest+

Wrong hashes could be a memory issue

**Comment by bigretromike** [comment]
@maxpiva thank you that you care about my setup but note needed.
just replace 1 character in file and you get same size different hash.

**Comment by da3dsoul** [comment]
I understand what he's saying. It's a problem that's not easily solved except by rehashing. @bigretromike there's a rehash all button in Unsorted in Nakamori.

The problem is:
- Two files with the same name and size. One is corrupted, likely a simple muxing error that results in identical filesize.
- when importing, it looks if we have a record with identical name and size, then uses that. This speeds up scanning moved files.
- Since we skip hashing, even though these files are different, they aren't treated as different. - They are assumed to be Duplicates, and the new one may be deleted if drop folders are used.
- If the new file is consistently deleted, then removing the old file is the only solution.

The only effective solution that I can think of is to rehash if the file we are replacing is unrecognized or manually linked. That should help.

**Comment by bigretromike** [comment]
just rehash didn't solve it. or maybe there is some kind of procedure to delete the old one somehow. 
when replacing old with new shoko assume is the same file and wont hash. 
and Rehash on server gui didn't act same way as rehash in unrecognized on client. 

Next time (hope I will never have next time) I will try it via Nakamori then, because Client worked very bad for me (it refreshed first page nonstop, but thats probably fault I use stable with daily via weak vpn tunel)

**Comment by maxpiva** [comment]
Figuring it out, Thomas, Do we always store filename/size and hash after
hashing for filename/size caching pruposes? If that so. We might need to
move that storage after identifying the file in AniDB instead after
hashing. That will prevent the filename/size db cache to become corrupt.

**Comment by da3dsoul** [comment]
Yes, it's there for a reason. It will properly overwrite if we rehash it, too. I think it's just a small logic issue that can be easily resolved. We don't want to change how it works, as it helps for situations with duplicates and relocating files. Right now, you can remap an entire import folder that's been moved around just by scanning it, and that's nice.

**Comment by bigretromike** [comment]
@da3dsoul probably the logic error you talk about there. 

also about `Thomas` when I get mail notification about your respond I see `Thomas` also

**Comment by da3dsoul** [comment]
What Rehash on server are you referring to? The Server tab in Desktop? The Server UI doesn't have a Rehash button

**Comment by da3dsoul** [comment]
@bigretromike?

**Comment by bigretromike** [comment]
Oh you right, Im refering to "import" - by clicking yellow icon next to Import Folder. 
I added file.mkv to folder /1/ that is in Import Folder.
It was corrupted.
I removed it from /1/ and added file.mkv to /1/new and hit Import.
I removed it from /1/new i moved it to /2/ and added it as ImportFolder and import it via yellow icon
still same thing
I also try with "remove missing files" from server thinking it will clean the file that I removed from /1/ but it didn't look like it worked then.

I had to ReHash them from Desktop Client.
Would be great to have that 'rehash' button on server also as its on server side anyway.

**Comment by da3dsoul** [comment]
Ah that only scans it. It doesn't force a rehash.
Yes the button should be in Server UI, but it isn't.

The idea I originally had causes too many problems. Namely, you're the only one who has had this issue and reported it. It's also theoretically improbable. Rehashing all unrecognized files on the improbable chance that they were hashed wrong or another file with the same name and size was imported with a different hash is just a waste of time and power. I often have 50GB BD releases that I won't keep end up in my Unrecognized Files because my downloader grabbed a remux. Rehashing those is pointless.

**Comment by da3dsoul** [comment]
I'll leave this open for now, as it does have some good points, like the need for a Rehash All Unrecognized Files button, but it's an edge case that really just needs user intervention.

**Comment by bigretromike** [comment]
In worst case just add button that query API for that if it's there already ;-) 

On March 28, 2018 8:52:09 AM GMT+02:00, da3dsoul <notifications@github.com> wrote:
>I'll leave this open for now, as it does have some good points, like
>the need for a Rehash All Unrecognized Files button, but it's an edge
>case that really just needs user intervention.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/ShokoAnime/ShokoServer/issues/724#issuecomment-376780966

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by da3dsoul** [comment]
I mean you can already do it in Nakamori per file. Just add a context item that calls:
api/rehashunlinked
api/rescanunlinked
To the raw files in Unsorted to do all unrecognized files.

**Comment by bigretromike** [comment]
I understand but it would be helpful in server UI because not everyone will use nakamori 

On March 28, 2018 10:22:19 AM GMT+02:00, da3dsoul <notifications@github.com> wrote:
>I mean you can already do it in Nakamori per file. Just add a context
>item that calls:
>api/rehashunlinked
>api/rescanunlinked
>To the raw files in Unsorted to do all unrecognized files.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/ShokoAnime/ShokoServer/issues/724#issuecomment-376802923

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by bigretromike** [comment]
@da3dsoul could we have that button on serverUI now or in 3.9 ?

**Comment by da3dsoul** [comment]
Maybe. It's not too hard

**Comment by da3dsoul** [comment]
I'll add it to the milestone and remove it if I decide I don't care enough atm

**Comment by ElementalCrisis** [comment]
Those actions are available in Desktop if you need to do it right now @bigretromike.

**Comment by bigretromike** [comment]
@ElementalCrisis I know, but its on server side anyway and API can do it
